nncf/dynamic_graph/context.py

```diff
@@ -13,24 +13,130 @@
 
 import re
 import threading
-from collections import deque
+from collections import deque, OrderedDict
 from contextlib import contextmanager
-from typing import Callable, List
+from functools import partial
+from typing import Callable, List, Dict, Tuple, Union
 from copy import deepcopy
+from itertools import islice
 
 from nncf.debug import is_debug
 from nncf.dynamic_graph.graph import InputAgnosticOperationExecutionContext
 from nncf.dynamic_graph.graph import NNCFGraph, NNCFNode
 from nncf.dynamic_graph.trace_tensor import make_input_infos
 from nncf.dynamic_graph.version_agnostic_op_names import get_version_agnostic_name
+from nncf.utils import maybe_get_iterator
 
 _CURRENT_CONTEXT = None
 
 
+def nth(iterable, n, default=None):
+    return next(islice(iterable, n, None), default)
+
+
+class InputIndexEntry:
+    def __init__(self, path: Tuple[Union[int, str], ...], getter: Callable, setter: Callable):
+        self.path = path
+        self.getter = getter
+        self.setter = setter
+
+class TupleRebuildingSetter:
+    def __init__(self, idx_to_set, current_tuple, previous_level_setter_for_current_tuple):
+        self._previous_level_setter = previous_level_setter_for_current_tuple
+        self._current_tuple = current_tuple
+        self._idx_to_set = idx_to_set
+
+    def __call__(self, value):
+        tmp_list = list(self._current_tuple)
+        tmp_list[self._idx_to_set] = value
+        new_tuple = tuple(tmp_list)
+        self._current_tuple = new_tuple
+        self._previous_level_setter(new_tuple)
+
+
 class OperatorInput:
```

Comment:
A PyTorch function input is not necessarily a single raw `torch.Tensor` for each input argument of the function. Consider `torch.cat([x1, x2])`, which has the input tensors `x1` and `x2` located inside a container, which itself is being passed as an argument. This function call, however, conceptually takes two tensors as its input, and if this is not traced properly, the NNCF graph will become disjoint. In a previous implementation this was circumvented by flattening the node inputs - a node's input signature, which identifies it among the other NNCF nodes, is a flat list of all tensors in the input arguments nesting hierarchy. 

The same approach is taken here - a flat view of the function's input tensors is being constructed, and each tensor is conceptually assigned a separate "input port ID" ordinal. However, this flat view must be a *view*, i.e. it should allow the tensors to be replaced with the results of pre-processing while preserving the original nested structure. Sadly, Python has no real notion of writeable iterators; therefore, it is necessary for each tensor in a flat view to store and use associated getter and setter functions, which are built using the closest owning container's `__getitem__` and `__setitem__` and the associated index of the tensor in the said container.